lib/wibox/widget/calendar.lua

```diff
@@ -0,0 +1,401 @@
+        forced_num_rows = 3,
+    }
+
+    local month_date
+    local current_day, current_month, current_year = parse_date(os.date("%d %m %Y"))
+
+    for month=1,12 do
+        if year == current_year and month == current_month then
+            month_date = string.format("%s %s %s", current_day, current_month, current_year)
+        else
+            month_date = string.format("%s %s", month, year)
+        end
+        in_layout:add(create_month(props, month_date))
+    end
+
+    -- Create a vertical layout
+    local flag, text = "yearheader", string.format("%s", year)
+    local year_header = props.fn_embed(make_cell(text, props.font, true), flag)
+    local out_layout = base.make_widget_declarative{
+        year_header,
+        in_layout,
+        spacing = 2*props.spacing, -- separate header from calendar grid
+        layout  = vertical
+    }
+    return props.fn_embed(out_layout, "year")
+end
+
+
+--- Set the container to the current date
+-- @param self Widget to update
+local function fill_container(self)
+    local date = self._private.date
+    if date then
+        -- Create calendar grid
+        if not self._private.calendar_cache[date] then
+            if self.type == "month" then
+                self._private.calendar_cache[date] = create_month(self._private, date)
+            elseif self.type == "year" then
+                self._private.calendar_cache[date] = create_year(self._private, date)
+            end
+        end
+        -- Create calendar box
+        self._private.container = self._private.calendar_cache[date]
+    else
+        self._private.container = nil
+    end
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Clear the cache and redraw
+-- @param self Widget to update
+local function clear_cache(self)
+    self._private.calendar_cache = {}
+    fill_container(self)
+end
+
+-- Set the calendar date
+function calendar:set_date(date)
+    if date ~= self._private.date then
+        self._private.date = date
+        -- Create calendar grid
+        fill_container(self)
+    end
+end
+
+
+-- Build properties function
+for _, prop in ipairs(properties) do
+    -- setter
+    if not calendar["set_" .. prop] then
+        calendar["set_" .. prop] = function(self, value)
+            if (string.sub(prop,1,3)=="fn_" and type(value) == "function") or self._private[prop] ~= value then
+                self._private[prop] = value
+                -- empty cache
+                clear_cache(self)
+                --self:emit_signal("widget::layout_changed")
+            end
+        end
+    end
+    -- getter
+    if not calendar["get_" .. prop] then
+        calendar["get_" .. prop] = function(self)
+            return self._private[prop]
+        end
+    end
+end
+
+-- Layout widget
+function calendar:layout(_, width, height)
+    if self._private.container then
+        return { base.place_widget_at(self._private.container, 0, 0, width, height) }
+    end
+    return {}
```

Comment:
This widget just wraps another widget. As I see it, that other widget is a `grid` layout. How about just creating the grid up-front and passing it as the first argument to `base.make_widget`, so that the `:fit` and `:layout` methods become unnecessary here.